basis/checksums/openssl/openssl.factor

```diff
@@ -38,27 +35,25 @@ M: evp-md-context dispose*
 : set-digest ( name ctx -- )
     handle>> swap digest-named f EVP_DigestInit_ex ssl-error ;
 
-: checksum-loop ( ctx -- )
-    dup handle>>
-    4096 read-partial dup [
-        dup length EVP_DigestUpdate ssl-error
-        checksum-loop
-    ] [ 3drop ] if ;
+: checksum-method ( data checksum method: ( ctx data -- ctx' ) -- value )
+    [ initialize-checksum-state ] dip '[ swap @ get-checksum ] with-disposal ; inline
+
+PRIVATE>
+
+M: openssl-checksum initialize-checksum-state ( checksum -- evp-md-context )
+    maybe-init-ssl name>> <evp-md-context> [ set-digest ] keep ;
 
-: digest-value ( ctx -- value )
+M: evp-md-context add-checksum-bytes ( ctx bytes -- ctx' )
+    [ dup handle>> ] dip dup length EVP_DigestUpdate ssl-error ;
+
+M: evp-md-context get-checksum ( ctx -- value )
     handle>>
     { { int EVP_MAX_MD_SIZE } int }
     [ EVP_DigestFinal_ex ssl-error ] with-out-parameters
     memory>byte-array ;
 
-PRIVATE>
+M: openssl-checksum checksum-bytes ( bytes checksum -- value )
+    [ add-checksum-bytes ] checksum-method ;
 
-M: openssl-checksum checksum-stream
-    name>> swap [
-        [
-            [ set-digest ]
-            [ checksum-loop ]
-            [ digest-value ]
-            tri
-        ] with-evp-md-context
-    ] with-input-stream ;
+M: openssl-checksum checksum-stream ( stream checksum -- value )
+    [ add-checksum-stream ] checksum-method ;
```

Comment:
I don't think it matters, my main point was to remove custom loop implementation in favor of the common code. I can benchmark.

As to whether to publish the checksum-method, it's your call, but it might be useful if all the checksums implement the common API pattern.